Extensible Enums, brought to you by Magic(TM) (#78)
neoforged/Documentation: docs/advanced/extensibleenums.md

:::warning
You should **not** be implementing this interface on your own enums. Use maps or registries instead depending on your usecase.
:::

### Creating an Enum Entry

To create a new enum entry, the static `create` method should be called. This static method is added to all extensible enums and is transformed at runtime to allow new enums to be added.

:::note
The `create` method should be called within a static final field if you need to refer to the value or within the main mod constructor if you do not.
:::

The `create` method contains a name parameter followed by the parameters of a supporting constructor. The name parameter represents the name of the enum constant, which is returned by `#name`. The name of the enum constant should be prefixed with your mod id followed by an underscore (`_`) to avoid conflicts between mods that decide to add the same enum constant (e.g., adding a enum constant called `test` with mod id `examplemod` should be `EXAMPLEMOD_TEST`).

:::note

> Since extensible enums are added at runtime, technically, any UTF-8 character can be used for the name paramter. However, it is recommended to only use valid Java identifiers.

Reviewer: ```suggestion
Since extensible enums are added at runtime, technically, any UTF-8 character can be used for the name parameter. However, it is recommended to only use valid Java identifiers.
```

:::

```java
// In your main mod class

> public static final FireworkExplosion.Shape DUMMY = FireworkExplosion.Shape.create("EXAMPLEMOD_DUMMY", 294, "examplemod_dummy");

Reviewer: ```suggestion
public static final FireworkExplosion.Shape DUMMY =
        FireworkExplosion.Shape.create("EXAMPLEMOD_DUMMY", 294, "examplemod_dummy");
```

:::note
The `create` method(s) will be transformed at runtime, so unless the enum was not detectable, the `create` method will create a new enum entry.
:::

```java
// This is an example, not an actual enum within Vanilla
public enum ExampleEnum implements net.neoforged.neoforge.common.IExtensibleEnum {
    // VALUE_1 represents the name parameter here
    VALUE_1(0, "value_1", false),
    VALUE_2(1, "value_2", true),
    VALUE_3(2, "value_3");

    ExampleEnum(int arg1, String arg2, boolean arg3) {
        // ...
    }

    ExampleEnum(int arg1, String arg2) {
        this(arg1, arg2, false);
    }

    // Matches the first constructor
    public static ExampleEnum create(String name, int arg1, String arg2, boolean arg3) {
        throw new IllegalStateException("Enum not extended");
    }

    // Matches the second constructor
    public static ExampleEnum create(String name, int arg1, String arg2) {
        throw new IllegalStateException("Enum not extended");
    }

    // ...
}
```

### The `init` method

> Sometimes, an enum will do some something with its constants or values after registration. However, as these instances statically call the associated entries, the values added via `create` may not be referenced. To get around this, extensible enums can override the `init` method to do any post-constructor setup required by the specific enum constant.

Reviewer: ```suggestion
Sometimes, an enum will do something with its constants or values after registration. However, as these instances statically call the associated entries, the values added via `create` may not be referenced. To get around this, extensible enums can override the `init` method to do any post-constructor setup required by the specific enum constant.
```

        Arrays.stream(ExampleEnumInit.values())
        .collect(
            Collectors.toMap(
                e -> e.arg2,
                e -> e.arg3
            )
        );

    ExampleEnumInit(int arg1, String arg2, boolean arg3) {
        // ...
    }
}

// The final extended enum may look something like this
public enum ExampleEnumInit implements net.neoforged.neoforge.common.IExtensibleEnum {
    VALUE_1(0, "value_1", false),
    VALUE_2(1, "value_2", true);

    private static final Map<String, boolean> ARG2_TO_ARG3 =
        Arrays.stream(ExampleEnumInit.values())
        .collect(
            Collectors.toMap(
                e -> e.arg2,
                e -> e.arg3
            )
        );

    ExampleEnumInit(int arg1, String arg2, boolean arg3) {
        // ...
    }

    // Matches the constructor
    public static ExampleEnumInit create(String name, int arg1, String arg2, boolean arg3) {
        throw new IllegalStateException("Enum not extended");
    }

    @Override
    public void init() {
        // Gets called for every new enum
        ARG2_TO_ARG3.put(this.arg2, this.arg3);
    }
}
```

### Codecs

> [Enum codecs][codec] generally take in the `values` array and resolves the codec instantly. As the array is evaluated before any mod entries are added, no mod entries will be supported. There are generally two solutions: which to choose depends on the current usecase.

Reviewer: ```suggestion
[Enum codecs][codec] generally take in the `values` array and resolve the codec instantly. As the array is evaluated before any mod entries are added, no mod entries will be supported. There are two solutions to this problem, which to choose depends on whether the enum implements the `StringRepresentable` interface or not.
```
